### DISCUSSION TITLE
How to load Steam dataset in python code like the ml-100k dataset?

### DISCUSSION THREAD

**Discussion by alanjacob10** [RUCAIBox/RecBole#792] [Q&A] [answered]
Hi,
I used the code:
from recbole.quick_start import run_recbole
run_recbole(dataset='ml-100k', model='BPR')

But was wondering if I could change the dataset name "ml-100k" to the Steam dataset you have on RecBole. What should I write instead of "ml-100k" to get the steam dataset? Does it have a specific name?

My goal is to test various algorithms on this dataset but I can't proceed because I can't load the Steam dataset.

ps. I'm relatively new to Python.

Sincerely,
Alan

**Comment by 2017pxy** [comment]
@alanjacob10 Hi, you need to download the steam dataset, and then change the config.

About the dataset, you can get it from [Our Google Drive](https://drive.google.com/drive/folders/1so0lckI6N6_niVEYaBu-LIcpOdZf99kj). Here I recommand you to create a folder call `dataset`  and you can put the steam dataset into this folder. The structure should be like:
```
-Dataset
    -steam
        -steam.inter
        -stram.item  
```

About the config setting, we support three way to set config(by command line, config files and parameter dicts). I will take the parameter dicts as an example:
```python
from recbole.quick_start import run_recbole

parameter_dict = {
    'data_path': "The file path of the Dataset"
    'model': BPR
    'dataset': steam
    'load_col’:
        inter: [ ]
    ......
}
run_recbole(config_dict=parameter_dict)
```  
For more infomation about our config settings, you can read our [docs](https://recbole.io/docs/user_guide/config_settings.html).

**Reply by alanjacob10** [reply]
Hi,
Thank you for the reply. I have downloaded the steam dataset from the drive but unfortunately I still have a problem with the code.

![Skærmbillede 2021-04-05 kl  14 44 36](https://user-images.githubusercontent.com/81938012/113575479-35c1a300-961e-11eb-803b-330fb51e26bf.png)
![Skærmbillede 2021-04-05 kl  14 44 44](https://user-images.githubusercontent.com/81938012/113575889-e0d25c80-961e-11eb-9e5b-45a0756e87eb.png)

It seems that BPR won't load. I also tried 
from recbole.model.general_recommender import BPR

but it didn't solve it.
I don't know if it will load the dataset. I used the "Steam_Merged" and set up the folders as you instructed.

ps. I'm using PyCharm and have installed recbole in the program.

Sincerely,
Alan

**Reply by 2017pxy** [reply]
@alanjacob10 Oh! I am sorry. I made a mistake. Actually, `model` and `dataset` can not be set in `parameter_dict`, and each parameter should be split by `,`. The correct code should be:
```python
from recbole.quick_start import run_recbole

parameter_dict = {
    'data_path': "The file path of the Dataset",
    'load_col':
        inter: [] #this should not be empty
    ......
}
run_recbole(model='BPR', dataset='steam' config_dict=parameter_dict)
```

By the way, the setting of `load_col` depends on the dataset and model, and it decides which column you want to load. For example, in `ml-100k` dataset, the `ml-100k.inter` has 4 columns (user_id,  item_id, rating, timestamp), if you only want to load `user_id` and `item_id`, the setting should be like:
```
 'load_col':
        inter: [user_id, item_id]
```
So I suggest you to check the header of `steam.inter` and set it by yourself.

Finally, sorry again for my carelessness. Please let me know if you have any question.

**Comment by alanjacob10** [comment]
Hi again,
Appreciate your kind response, but unfortunately I'm still struggling with the code. I did as you instructed and also found the column names from the "steam_inter" file, where I input user_id, product_id. 
And here was the result:

![Skærmbillede 2021-04-06 kl  01 14 56](https://user-images.githubusercontent.com/81938012/113637957-10608380-9676-11eb-8180-ee0ca9152fbf.png)
![Skærmbillede 2021-04-06 kl  01 15 19](https://user-images.githubusercontent.com/81938012/113637959-1191b080-9676-11eb-8f16-216926442d9c.png)

I don't know if i need to import something else from Pycharm other than recbole or what the problem is.

Sincerely,
Alan

**Reply by 2017pxy** [reply]
@alanjacob10 Oh! I am so so sorry. The dict should be:
```
parameter_dict = {
    'data_path': "The file path of the Dataset",
    'load_col':  {'inter': ['user_id', 'product_id']}
}
```
Please try again and forgive me for my carelessness again.
(Oh, it's too embarrassed)

**Reply by alanjacob10** [reply]
Hi @2017pxy 

The parameter_dict seems to be working now, but there is a problem with loading the file. It gives the following error.

![Skærmbillede 2021-04-06 kl  12 04 20](https://user-images.githubusercontent.com/81938012/113694998-ade8a100-96d0-11eb-96ac-9b35b109e3a2.png)
![Skærmbillede 2021-04-06 kl  12 04 37](https://user-images.githubusercontent.com/81938012/113695006-afb26480-96d0-11eb-8287-65251b90889e.png)

ps. I also tried switching to another dataset from your drive to test if it was working with a different .inter file.

Sincerely,
Alan

**Reply by hyp1231** [reply]
Hi Alan, the value of your param "data_path" should be the path of "dataset" dir, while not the path of "xxx.inter".

**Reply by alanjacob10** [reply]
Hi,
@hyp1231 I tried this also, but it also gave an error. see the following:

![Skærmbillede 2021-04-06 kl  12 40 22](https://user-images.githubusercontent.com/81938012/113699080-69abcf80-96d5-11eb-9db7-8ec41b202f2a.png)


Sincerely,
Alan

**Reply by hyp1231** [reply]
Hi Alan,
could you please show the file structure of your "dataset" dir? Thanks.

**Reply by alanjacob10** [reply]
yes, it is as follows:

-dataset
    -steam
        -steam.inter
        -stram.item 

Thanks,
Alan

**Reply by tsotfsk** [reply]
It seems that the message `*/dataset/steam/steam` in error is different form what you set. Do you set up any other config files?

**Reply by alanjacob10** [reply]
Hi,
I don't know any config files. It should be the only one set up. Do you know this error? it mentions pandas...
![Skærmbillede 2021-04-06 kl  13 39 16](https://user-images.githubusercontent.com/81938012/113707092-b3011c80-96df-11eb-8db1-ce7163e35ba4.png)

Regards,
Alan

**Reply by alanjacob10** [reply]
@2017pxy

**Reply by tsotfsk** [reply]
Hi ! @alanjacob10 ,  Maybe you can set up a config file named `steam.yaml`  which contains the following three lines
```yaml
load_col:  {inter: [user_id, product_id]}
USER_ID_FIELD: user_id
ITEM_ID_FIELD: product_id
```
Then try run your model by
```shell
python run_recbole.py --config_files steam.yaml --dataset steam --model BPR
```

**Comment by alanjacob10** [comment]
?

**Reply by 2017pxy** [reply]
@alanjacob10 Hi, sorry for late reply. I install the Pycharm on Windows and successfully run the BPR model in steam dataset with the parameter_dict as:
```
parameter_dict = {
    'data_path': "YOUR PATH",
    'load_col': {'inter': ['user_id', 'product_id']},
    'USER_ID_FIELD': 'user_id',
    'ITEM_ID_FIELD': 'product_id'
}
```
I must admit that there are lots of config settings in RecBole and my suggestion is: read our [docs](https://recbole.io/docs/) first. And we will consider adding more user-friendly examples in our docs in near future.

**Reply by alanjacob10** [reply]
Hi @2017pxy ,
This is very good. Thank you for your great solution and help!
Might I ask if I were to add the parameter play_hours from steam dataset (I don't know on which algorithm would be suitable), but how would this extra parameter of play_hours be added to your line of code above?
would it be:
'PLAY_HOURS_FIELD': 'play_hours'

Or maybe this is a 'RATING_FIELD' instead?

Sincerely,
Alan

**Reply by 2017pxy** [reply]
@alanjacob10 Hi, if you want to load the column from you file, you only need to change your `load_col`. We don't have `PLAY_HOURS_FIELD` in our config settings. For example, if you want to load `play_hours`, your  `load_col` should be:
```
'load_col': {'inter': ['user_id', 'product_id', 'play_hours']}
```
By the way, for general recommendation models like BPR, they will only use `user_id` and `item_id` in training phase. That's why `USER_ID_FIELD` and `ITEM_ID_FIELD` should be set base on your dataset. You need to tell the model which columns are `USER_ID_FIELD` and `ITEM_ID_FIELD`. 

For other information, like `play_hours` in steam dataset, is viewed as context information. If you want to use context information to train the models in RecBole, you should use context-aware recommendation models like [LR](https://recbole.io/docs/user_guide/model/context/lr.html) and [FM](https://recbole.io/docs/user_guide/model/context/fm.html). For more information about the models supported in RecBole, please check our [model docs](https://recbole.io/docs/user_guide/model_intro.html#context-aware-recommendation).

**Reply by alanjacob10** [reply]
One follow-up question to this.. According to your documentation, context-aware recommendation models requires: 
.inter , .user and .item file. 
Does all three files need to be there to run the algorithm??

And how do I add these to load col. ?

from recbole.quick_start import run_recbole

parameter_dict = {
    'data_path': "/Users/alanolewnik/Documents/PyCharm/Recommender_new/dataset",
    'load_col': {'inter': ['user_id', 'product_id', 'timestamp', 'play_hours']},
    'USER_ID_FIELD': 'user_id',
    'ITEM_ID_FIELD': 'product_id'
}

run_recbole(model='DeepFM', dataset='steam', config_dict=parameter_dict)



Sincerely,
Alan

**Reply by hyp1231** [reply]
> One follow-up question to this.. According to your documentation, context-aware recommendation models requires:
> .inter , .user and .item file.
> Does all three files need to be there to run the algorithm??
> 
> And how do I add these to load col. ?
> 
> from recbole.quick_start import run_recbole
> 
> parameter_dict = {
> 'data_path': "/Users/alanolewnik/Documents/PyCharm/Recommender_new/dataset",
> 'load_col': {'inter': ['user_id', 'product_id', 'timestamp', 'play_hours']},
> 'USER_ID_FIELD': 'user_id',
> 'ITEM_ID_FIELD': 'product_id'
> }
> 
> run_recbole(model='DeepFM', dataset='steam', config_dict=parameter_dict)
> 
> Sincerely,
> Alan

Hi Alan, to run a context-aware recommendation model, `.inter` is required, while `.user` and `.item` are optional.

Could you please point out the detail urls about the misleading docs? So that we will fix the descriptions, thanks!

**Reply by alanjacob10** [reply]
Okay thanks. It says in this box.. (see attached):
![2021-04-22 172546](https://user-images.githubusercontent.com/81938012/115741303-e7d3cb80-a38f-11eb-9e74-0747f3d76b55.png)
https://recbole.io/docs/user_guide/data/atomic_files.html

Sincerely,
Alan

**Reply by hyp1231** [reply]
Got it, thanks so much. It's indeed misleading description. We'll consider improving this part.

**Comment by 2017pxy** [comment, marked as answer]
Let me make a summary to this question.
If you want to load a new dataset to run models, you can follow these steps:
1. Prepare your dataset files:
In RecBole, we have a default dataset: ml-100k. If you want to use other dataset, you need to prepare your data and convert the raw data into `Atomic Files`(About Atomic Files, here is the [docs](https://recbole.io/docs/user_guide/data/atomic_files.html)). By the way, we have prepared some [popular datasets](https://recbole.io/dataset_list.html) and you can download the atomic files of these datasets from our [Google Drive](https://drive.google.com/drive/folders/1so0lckI6N6_niVEYaBu-LIcpOdZf99kj) or [Baidu Wangpan](https://pan.baidu.com/s/1p51sWMgVFbAaHQmL4aD_-g#list/path=%2F). Then, create a folder called `MyDataset` and organize the file structure like:
```
-MyDataset
    -DataA
        -DataA.inter
        -DataA.item
        - ......
    -DataB
        -DataB.inter
    ........
```
  
2. Set config:
If you load a new dataset, the default config settings may need to be changed, so you need to reset the config by yourself. Before you do this, I strongely recommend you to read our [config setting docs](https://recbole.io/docs/user_guide/config_settings.html) and [data args docs](https://recbole.io/docs/user_guide/data/data_args.html) first, or you may face lots of problems. 
About the config setting, here are some settings you may need to change:
```yaml
data_path: The path of "MyDataset" (mentioned before)
load_col: Decide which file and column you want to load;
USER_ID_FIELD: Field name of user ID feature
ITEM_ID_FIELD: Field name of item ID feature
RATING_FIELD: Field name of rating feature
TIME_FIELD: Field name of timestamp feature
```

3. Load settings and run the model:
In RecBole, we support three way to load settings: config files, parameter dicts and command line. You can read our [config setting docs](https://recbole.io/docs/user_guide/config_settings.html)  to find the details and examples about loading the settings. And then, you can simply run the model with RecBole and finish your research.

**Reply by thanh31596** [reply]
it got so many weird bugs

**Comment by Faisalse** [comment]
Hi Guys,
In the RecBole framework, can we use already train-test split files to tune or train the models?
